File: src/main/java/dev/raphaeldelio/redisapollo/photograph/ImageController.java
package dev.raphaeldelio.redisapollo.photograph;

import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
@RequestMapping("/image")
public class ImageController {

    private final ImageService imageService;

    public ImageController(ImageService imageService) {
        this.imageService = imageService;
    }

    @PostMapping("/search/by-image")
    public Map<String, Object> searchByImage(@RequestBody ImageSearchRequest request) {
        long start = System.currentTimeMillis();
        String tmpPath = imageService.saveTmpImage(request.imageBase64(), request.imagePath());
        return Map.of(
                "imagePath", tmpPath,
                "matchedPhotographs", imageService.searchByImage(tmpPath),
                "processingTime", (System.currentTimeMillis() - start) + "ms"
        );
    }

    @PostMapping("/search/by-description")
    public Map<String, Object> searchByImageText(@RequestBody SearchRequest request) {
        long start = System.currentTimeMillis();
        return Map.of(
                "query", request.query(),
                "matchedPhotographs", imageService.searchByImageText(request.query()),
                "processingTime", (System.currentTimeMillis() - start) + "ms"
        );
    }

    // ---------- DTOs ----------

    public record SearchRequest(String query, boolean enableSemanticCache, boolean enableRag) {}

    public record ImageSearchRequest(String imageBase64, String imagePath) {}
}

File: redis-apollo-app-kotlin/src/main/kotlin/dev/raphaeldelio/redisapollo/photograph/ImageController.kt
package dev.raphaeldelio.redisapollo.photograph

import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/image")
class ImageController(
    private val imageService: ImageService
) {

    @PostMapping("/search/by-image")
    fun searchByImage(@RequestBody request: ImageSearchRequest): Map<String, Any> {
        val tmpPath = imageService.saveTmpImage(request.imageBase64, request.imagePath)

        val embeddingStart = System.currentTimeMillis()
        val tmpPhoto = imageService.embedImage(tmpPath)
        val embeddingTime = System.currentTimeMillis() - embeddingStart

        val searchStart = System.currentTimeMillis()
        val matchedPhotographs = imageService.searchByImage(tmpPhoto.embeddedImage!!)
        val searchTime = System.currentTimeMillis() - searchStart

        return mapOf(
            "imagePath" to tmpPath,
            "matchedPhotographs" to matchedPhotographs,
            "embeddingTime" to "${embeddingTime}ms",
            "searchTime" to "${searchTime}ms"
        )
    }

    @PostMapping("/search/by-description")
    fun searchByImageText(@RequestBody request: SearchRequest): Map<String, Any> {
        val embeddingStart = System.currentTimeMillis()
        val embedding = imageService.embedDescription(request.query)
        val embeddingTime = System.currentTimeMillis() - embeddingStart

        val searchStart = System.currentTimeMillis()
        val matchedPhotographs = imageService.searchByImageText(embedding)
        val searchTime = System.currentTimeMillis() - searchStart

        return mapOf(
            "query" to request.query,
            "matchedPhotographs" to matchedPhotographs,
            "embeddingTime" to "${embeddingTime}ms",
            "searchTime" to "${searchTime}ms"
        )
    }

    data class SearchRequest(
        val query: String,
        val enableSemanticCache: Boolean,
        val enableRag: Boolean
    )

    data class ImageSearchRequest(
        val imageBase64: String,
        val imagePath: String?
    )
}